File: haskell/src/Day02.hs
module Day02 where

import Data.List.Split

type Password = String

data Policy = Policy Int Int Char deriving (Show)

inputDecode :: [String] -> [(Policy, Password)]
inputDecode [] = []
inputDecode (x : xs) =
  let [policy, password] = splitOn ":" x
      [range, char] = words policy
      [atleast, atmost] = map read . splitOn "-" $ range
   in (Policy atleast atmost (head char), head . words $ password) : inputDecode xs

count :: (a -> Bool) -> [a] -> Int
count f = length . filter f

xorBool :: Bool -> Bool -> Bool
xorBool a b = (a && not b) || (not a && b)

part1 :: [(Policy, Password)] -> Int
part1 = count (\(Policy lo hi c, pass) -> count (== c) pass <= hi && count (== c) pass >= lo)

part2 :: [(Policy, Password)] -> Int
part2 = count (\(Policy p1 p2 c, pass) -> xorBool (pass !! (p1 - 1) == c) (pass !! (p2 -1) == c))

File: elixir/lib/advent_of_code/day_02.ex
defmodule AdventOfCode.Day02 do
  def input_decode(inputs) do
    inputs
    |> Enum.map(&String.split(&1, ": "))
    |> Enum.map(fn [pol, pass] ->
      [r, char] = String.split(pol)
      [lo, hi] = String.split(r, "-") |> Enum.map(&String.to_integer/1)
      {lo, hi, char, pass}
    end)
  end

  def count(fun, list) do
    length(Enum.filter(list, fun))
  end

  def xor_bool(a, b) do
    (a and not b) or (not a and b)
  end

  def part1(list) do
    count(
      fn {lo, hi, c, pass} ->
        count(&(&1 === hd(to_charlist(c))), to_charlist(pass)) in lo..hi
      end,
      list
    )
  end

  def part2(list) do
    count(
      fn {p1, p2, c, pass} ->
        xor_bool(String.at(pass, p1 - 1) == c, String.at(pass, p2 - 1) == c)
      end,
      list
    )
  end
end
